src/cairoWriter/writers/cairoContractWriter.ts

```diff
@@ -98,19 +100,84 @@ export class CairoContractWriter extends CairoASTNodeWriter {
       }
     `;
 
+    // Data about imports used, util funcs, constants and structs definition are stored in the Source Unit node
+    // but should be printed inside the contract module
+
+    // Only constants generated by `newToDeploy` exist at this stage
+    const constants = sourceUnit.vVariables
+      .map((v) => {
+        assert(v.vValue !== undefined, 'Constants cannot be unanssigned');
+        return [`// ${v.documentation}`, `const ${v.name} = ${writer.write(v.vValue)};`].join('\n');
+      })
+      .join('\n');
+
+    const freeStructs = getStructs(sourceUnit, this.ast);
+    const structs = [...freeStructs, ...sourceUnit.vStructs, ...(node?.vStructs || [])]
+      .map((v) => writer.write(v))
+      .join('\n\n');
+
+    const importFunctions = sourceUnit.vFunctions.filter(
+      (f): f is CairoImportFunctionDefinition => f instanceof CairoImportFunctionDefinition,
+    );
+    const generatedFunctions = sourceUnit.vFunctions.filter(
+      (f): f is CairoGeneratedFunctionDefinition => f instanceof CairoGeneratedFunctionDefinition,
+    );
+    const otherSourceUnitFunctions = sourceUnit.vFunctions.filter(
+      (f) =>
+        !(f instanceof CairoGeneratedFunctionDefinition) &&
+        !(f instanceof CairoImportFunctionDefinition),
+    );
+
+    const writtenImportFuncs = importFunctions
+      .map((n) => writer.write(n))
+      .sort((importA, importB) => importA.localeCompare(importB))
+      .filter((func, index, importFuncs) => func !== importFuncs[index - 1])
+      .join('\n');
+
+    const writtenGeneratedFuncs = generatedFunctions
+      .sort((funcA, funcB) => funcA.name.localeCompare(funcB.name))
+      .sort((funcA, funcB) => {
+        const stubA = funcA.functionStubKind;
+        const stubB = funcB.functionStubKind;
+        if (stubA === stubB) return 0;
+        if (stubA === FunctionStubKind.StructDefStub) return -1;
+        if (stubA === FunctionStubKind.StorageDefStub) return -1;
+        return 1;
+      })
+      .filter(
+        (func, index, genFuncs) =>
+          func.name !== genFuncs[index - 1]?.name &&
+          func.functionStubKind !== FunctionStubKind.StorageDefStub,
+      )
+      .map((func) => writer.write(func))
+      .join('\n\n');
+
+    const writtenOtherSourceUnitFunctions = otherSourceUnitFunctions
+      .map((func) => writer.write(func))
+      .join('\n\n');
+
+    const contractHeader = '#[contract] \n' + `mod ${node.name} {`;
+
     return [
       [
+        contractHeader,
         documentation,
-        ...events,
-        `namespace ${node.name}{\n\n${body}\n\n}`,
-        outsideNamespaceBody,
+        writtenImportFuncs,
```

Comment:
Shouldn't the written importFuncs be on the top level of the script?

Reply:
Yes it seems module related imports are defined inside the module. I agree with leaving them here

Reply:
I saw this other [example](https://github.com/starkware-libs/cairo/blob/main/crates/cairo-lang-starknet/test_data/account.cairo) where they use imports outside of the module as well, at the top of the file as you had suggested; but I think this is something else  we can work on later. I made a new issue to address this task and explain it better https://github.com/NethermindEth/warp/issues/997